x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/create_package_policy_page/hooks/navigation.tsx

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { useCallback, useMemo, useEffect, useRef } from 'react';
+import { useHistory } from 'react-router-dom';
+
+import { useStartServices, useLink, useIntraAppState } from '../../../../hooks';
+import type {
+  CreatePackagePolicyRouteState,
+  PackagePolicy,
+  NewPackagePolicy,
+  OnSaveQueryParamKeys,
+} from '../../../../types';
+import type { EditPackagePolicyFrom } from '../types';
+
+import { appendOnSaveQueryParamsToPath } from '../utils';
+
+interface UseCancelParams {
+  from: EditPackagePolicyFrom;
+  pkgkey: string;
+  agentPolicyId?: string;
+}
+
+export const useCancelAddPackagePolicy = (params: UseCancelParams) => {
+  const { from, pkgkey, agentPolicyId } = params;
+  const {
+    application: { navigateToApp },
+  } = useStartServices();
+  const routeState = useIntraAppState<CreatePackagePolicyRouteState>();
+  const { getHref } = useLink();
+
+  const cancelClickHandler = useCallback(
+    (ev) => {
+      if (routeState?.onCancelNavigateTo) {
+        ev.preventDefault();
+        navigateToApp(...routeState.onCancelNavigateTo);
+      }
+    },
+    [routeState, navigateToApp]
+  );
+
+  const cancelUrl = useMemo(() => {
+    if (routeState && routeState.onCancelUrl) {
+      return routeState.onCancelUrl;
+    }
+    return from === 'policy' && agentPolicyId
+      ? getHref('policy_details', {
+          policyId: agentPolicyId,
+        })
+      : getHref('integration_details_overview', { pkgkey });
+  }, [routeState, from, agentPolicyId, getHref, pkgkey]);
+
+  return { cancelClickHandler, cancelUrl };
+};
+
+interface UseOnSaveNavigateParams {
+  packagePolicy: NewPackagePolicy;
+  routeState?: CreatePackagePolicyRouteState;
+  queryParamsPolicyId?: string;
+}
+
+export const useOnSaveNavigate = (params: UseOnSaveNavigateParams) => {
+  const { packagePolicy, queryParamsPolicyId } = params;
+  const routeState = useIntraAppState<CreatePackagePolicyRouteState>();
+  const doOnSaveNavigation = useRef<boolean>(true);
+  const { getPath } = useLink();
+  const history = useHistory();
+
+  const {
+    application: { navigateToApp },
+  } = useStartServices();
+
+  // Detect if user left page
+  useEffect(() => {
+    return () => {
+      doOnSaveNavigation.current = false;
+    };
+  }, []);
```

Comment:
This feels like it could be a problem if the `params` change, but I haven't thought through all the implications.